sdks/go/pkg/beam/core/runtime/graphx/schema/schema.go

```diff
@@ -0,0 +1,245 @@
+		isPtr = true
+		t = t.Elem()
+	}
+	switch t.Kind() {
+	case reflect.Map:
+		kt := reflectTypeToFieldType(t.Key())
+		vt := reflectTypeToFieldType(t.Elem())
+		return &pipepb.FieldType{
+			Nullable: isPtr,
+			TypeInfo: &pipepb.FieldType_MapType{
+				MapType: &pipepb.MapType{
+					KeyType:   kt,
+					ValueType: vt,
+				},
+			},
+		}
+	case reflect.Struct:
+		sch := structToSchema(t)
+		return &pipepb.FieldType{
+			Nullable: isPtr,
+			TypeInfo: &pipepb.FieldType_RowType{
+				RowType: &pipepb.RowType{
+					Schema: sch,
+				},
+			},
+		}
+	case reflect.Slice, reflect.Array:
+		// Special handling for []byte
+		if t == reflectx.ByteSlice {
+			return &pipepb.FieldType{
+				Nullable: isPtr,
+				TypeInfo: &pipepb.FieldType_AtomicType{
+					AtomicType: pipepb.AtomicType_BYTES,
+				},
+			}
+		}
+		vt := reflectTypeToFieldType(t.Elem())
+		return &pipepb.FieldType{
+			Nullable: isPtr,
+			TypeInfo: &pipepb.FieldType_ArrayType{
+				ArrayType: &pipepb.ArrayType{
+					ElementType: vt,
+				},
+			},
+		}
+	case reflect.Interface, reflect.Chan, reflect.UnsafePointer, reflect.Complex128, reflect.Complex64, reflect.Int:
+		panic(fmt.Sprintf("Unsupported type to convert to schema: %v", ot))
+	default: // must be an atomic type
+		enum := reflectTypeToAtomicType(t)
+		return &pipepb.FieldType{
+			Nullable: isPtr,
+			TypeInfo: &pipepb.FieldType_AtomicType{
+				AtomicType: enum,
+			},
+		}
+	}
+}
+
+func reflectTypeToAtomicType(rt reflect.Type) pipepb.AtomicType {
+	switch rt {
+	case reflectx.Uint8:
+		return pipepb.AtomicType_BYTE
+	case reflectx.Int16:
+		return pipepb.AtomicType_INT16
+	case reflectx.Int32:
+		return pipepb.AtomicType_INT32
+	case reflectx.Int64, reflectx.Int:
+		return pipepb.AtomicType_INT64
+	case reflectx.Float32:
+		return pipepb.AtomicType_FLOAT
+	case reflectx.Float64:
+		return pipepb.AtomicType_DOUBLE
+	case reflectx.String:
+		return pipepb.AtomicType_STRING
+	case reflectx.Bool:
+		return pipepb.AtomicType_BOOLEAN
+	case reflectx.ByteSlice:
+		return pipepb.AtomicType_BYTES
+	default:
+		panic(fmt.Sprintf("non atomic reflect type: %v", rt))
+	}
+}
+
+// ToType returns a Go type of the passed in Schema.
+// Types returned by ToType are always of Struct kind.
+// Returns an error if the Schema cannot be converted to a type.
+func ToType(s *pipepb.Schema) (reflect.Type, error) {
+	fields := make([]reflect.StructField, 0, len(s.GetFields()))
+	for _, sf := range s.GetFields() {
+		rf := fieldToStructField(sf)
+		fields = append(fields, rf)
+	}
+	return reflect.StructOf(fields), nil
+}
+
+func fieldToStructField(sf *pipepb.Field) reflect.StructField {
+	name := sf.GetName()
+	return reflect.StructField{
+		Name: strings.ToUpper(name[:1]) + name[1:], // Go field name must be capitalized for export and encoding.
+		Type: fieldTypeToReflectType(sf.GetType()),
+		Tag:  reflect.StructTag(fmt.Sprintf("beam:\"%s\"", name)),
+	}
+}
+
+var atomicTypeToReflectType = map[pipepb.AtomicType]reflect.Type{
```

Comment:
Using a map for decoding atomic types and a switch statement for encoding them seems inconsistent. Is there a reason for it? If not I'd say to go with one approach in both cases.

Reply:
Honestly, when I first wrote it, it complained that reflect.Type couldn't be used as a key, but I do that all the time so... very strange.